### PR TITLE
#1355 Migration tool

### PR DESCRIPTION
**Pull request by lizonpatryk**
related: #1355 
**DONE**
- [x]  Migration Tool Smart Contract Logic
- [x]  Initial Visuals
- [x]  Wallet Periodic Updating
- [x]  Fixes GasPrice Api

**TODO:**
- [ ] Add myEtherWalletLink
- [ ] Visual Fixes

### CODE REVIEW

File: app/modules/icbmWalletBalanceModal/sagas.ts [hunk 1]
```diff
@@ -1,25 +1,80 @@
...
+          gasLimit: "400000",
+          value: "0",
+        }
+      : {
+          migrationInputData: contractsService.etherLock
+            .setInvestorMigrationWalletTx(currentEthAddress)
+            .getData(),
+          migrationStep: 1,
+          smartContractAddress: etherLockAddress,
+          gasLimit: "200000",
```
Comment (rudolfix) [inline app/modules/icbmWalletBalanceModal/sagas.ts]: Please set gas limit to 400000, just checked that 200000 does not work in some cases!

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      .click();
+    cy.get(tid("modals.icbm-balance-modal.migrate-body.to")).then(toField => {
+      const to = toField.text().replace(charRegExPattern, "");
+
+      cy.get(tid("modals.icbm-balance-modal.migrate-body.gas-limit")).then(gasLimitField => {
+        const gas = gasLimitField.text().replace(charRegExPattern, "");
+
+        cy.get(tid("modals.icbm-balance-modal.migrate-body.input-data")).then(inputDataField => {
+          const data = inputDataField.text().replace(charRegExPattern, "");
+          getNonceRpc(NODE_ADDRESS, "0x429123b08DF32b0006fd1F3b0Ef893A8993802f3").then(nonce => {
```
Comment (rudolfix) [inline app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: I would promisify all of this

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                  nonce: nonce.body.result,
+                  chainId: chainId.body.result,
+                })
+                .then((signed: any) => {
+                  cy.log("Sending First Transaction");
+                  sendRawTransactionRpc(NODE_ADDRESS, signed.rawTransaction).then(hash => {
+                    // Wait for transaction to get conducted
+                    cy.wait(1000);
+                    getTransactionReceipt(NODE_ADDRESS, hash.body.result).then(receipt => {
+                      cy.get(tid("modals.icbm-balance-modal.migrate-body.step-2"));
```
Comment (rudolfix) [inline app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: should we at least check `status` == 1 in transaction receipt?

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                                          NODE_ADDRESS,
+                                          signed.rawTransaction,
+                                        ).then(hash => {
+                                          // Wait for transaction to get conducted
+                                          cy.wait(1000);
+                                          getTransactionReceipt(
+                                            NODE_ADDRESS,
+                                            hash.body.result,
+                                          ).then(receipt => {
+                                            console.log(receipt);
```
Comment (rudolfix) [inline app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: I'd also check status explicitely here

File: app/modules/icbmWalletBalanceModal/sagas.ts [hunk 5]
```diff
@@ -1,25 +1,80 @@
...
+    );
+    const investorMigrationWallet: [
+      string[],
+      BigNumber[]
+    ] = yield contractsService.etherLock.getInvestorMigrationWallets(icbmEthAddress);
+
+    const didUserConductFirstTransaction = !!(
+      investorMigrationWallet[0][0] &&
+      investorMigrationWallet[0][0].toLowerCase() === currentEthAddress.toLowerCase() &&
+      investorMigrationWallet[0].length === 1
```
[the remark this answers is not in the document]
Comment (Moejoe90) [reply app/modules/icbmWalletBalanceModal/sagas.ts]: Actually, you have a point it should be on top of the boolean expression. 
In all cases, I was under the impression that a user can manipulate values and have more than one address in this array by adding more than one address or so. 
So this check is to make sure the user is following a clear path.

File: app/modules/icbmWalletBalanceModal/sagas.ts [hunk 6]
```diff
@@ -52,6 +108,42 @@ function* loadIcbmWalletDataSaga(
...
+
+  while (true) {
+    const investorMigrationWallet: [
+      string[],
+      BigNumber[]
+    ] = yield contractsService.etherLock.getInvestorMigrationWallets(icbmEthAddress);
+    const didUserConductFirstTransaction = !!(
+      investorMigrationWallet[0][0] &&
+      investorMigrationWallet[0][0].toLowerCase() === currentEthAddress.toLowerCase() &&
+      investorMigrationWallet[0].length === 1
```
[the remark this answers is not in the document]
Comment (Moejoe90) [reply app/modules/icbmWalletBalanceModal/sagas.ts]: done

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+const NODE_ADDRESS = "https://localhost:9090/node";
+
+describe("Auto Login", () => {
+  beforeEach(() =>
+    createAndLoginNewUser({
+      type: "investor",
+      kyc: "individual",
+    }));
+
+  it("will auto login", () => {
```
[the remark this answers is not in the document]
Comment (Moejoe90) [reply app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: :+1:

File: app/modules/icbmWalletBalanceModal/sagas.ts (hunk 1 above)
Comment (Moejoe90) [reply app/modules/icbmWalletBalanceModal/sagas.ts]: done

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts (hunk 2 above)
Comment (Moejoe90) [reply app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: Its a bit tricky cypress doesn't fully support promises in the way regular way but there are a couple of things that can be done to make that possible and have a nice `await/async` clean flow

https://github.com/cypress-io/cypress/issues/1417

I have plans refactor this in the future and add support for regular promises using for example
https://github.com/NicholasBoll/cypress-promise

For now i will focus on other critical tickets

File: app/test-e2e/wallets/WalletMigrate.spec.e2e.ts (hunk 4 above)
Comment (Moejoe90) [reply app/test-e2e/wallets/WalletMigrate.spec.e2e.ts]: Done

**Comment by Moejoe90** [comment]
image
Current Message

**Comment by Moejoe90** [comment]
image

**Comment by Moejoe90** [comment]
image
Current Message

**Pull request merged**